### PR TITLE
Add ability to filter triton_sd targets by pre-defined groups

### PR DESCRIPTION
**Pull request by richardkiene**
Signed-off-by: Richard Kiene <richard.kiene@joyent.com>

As Triton and Manta installs have become substantially larger and customer reliance upon Prometheus has become more important, it has become painfully obvious that filtering is a necessary feature of Triton Service Discovery. This pull request leverages the newly added group filtering support that we've added to [Triton CMON](https://github.com/joyent/triton-cmon).

//cc: @fabxc @brian-brazil

### CODE REVIEW

**Comment by brian-brazil** [comment]
I don't see the groups among the current metadata, usually we use relabelling to do filtering rather than having every SD implement it's own ad-hoc mechanisms. Could you add this information to the metadata returned and work from that?

**Comment by brian-brazil** [comment]
> that is already possible with other metadata, but at this point we're throwing away 10's of thousands of targets that come back from the endpoint with that method.

I don't see the group among the current data. Is this actually causing performance problems, or is it more theoretical?

**Comment by richardkiene** [comment]
@brian-brazil Yes it is actually causing performance problems. Groups were very recently added to support filtering in this exact way, so I'm not sure why they would also be in the metadata.

**Comment by brian-brazil** [comment]
Can you add in the metadata in this change then also please?

I'd also suggest waiting until the support in triton-cmon is actually merged.

**Comment by richardkiene** [comment]
@brian-brazil I've updated the PR to include support for groups in metadata. The corresponding changes for triton-cmon will be merged to master sometime during business hours on Monday Oct 8th 2018 Pacific Time. If you'd like to track that process you can watch https://cr.joyent.us/#/c/4888/. That said, I've tested these changes against an older version of triton-cmon and they should not break users who are using an older versions.

File: discovery/triton/triton.go [hunk 1]
```diff
@@ -102,11 +105,12 @@ func init() {
 // DiscoveryResponse models a JSON response from the Triton discovery.
 type DiscoveryResponse struct {
 	Containers []struct {
-		ServerUUID  string `json:"server_uuid"`
-		VMAlias     string `json:"vm_alias"`
-		VMBrand     string `json:"vm_brand"`
-		VMImageUUID string `json:"vm_image_uuid"`
-		VMUUID      string `json:"vm_uuid"`
+		Groups      []string `json:"groups:`
```
Comment (brian-brazil) [inline discovery/triton/triton.go]: The tag is off here

File: discovery/triton/triton.go [hunk 2]
```diff
@@ -187,6 +191,10 @@ func (d *Discovery) refresh() (tg *targetgroup.Group, err error) {
 	}()
 
 	var endpoint = fmt.Sprintf("https://%s:%d/v%d/discover", d.sdConfig.Endpoint, d.sdConfig.Port, d.sdConfig.Version)
+	if d.sdConfig.Groups != nil {
+		endpoint = fmt.Sprintf("%s?groups=%s", endpoint, strings.Join(d.sdConfig.Groups, ","))
```
Comment (brian-brazil) [inline discovery/triton/triton.go]: Is there any escaping needed here?

File: discovery/triton/triton.go [hunk 3]
```diff
@@ -219,6 +227,13 @@ func (d *Discovery) refresh() (tg *targetgroup.Group, err error) {
 		}
 		addr := fmt.Sprintf("%s.%s:%d", container.VMUUID, d.sdConfig.DNSSuffix, d.sdConfig.Port)
 		labels[model.AddressLabel] = model.LabelValue(addr)
+
+		if container.Groups != nil {
+			for _, v := range container.Groups {
+				labels[tritonLabelGroups+"_"+model.LabelName(v)] = model.LabelValue(v)
```
Comment (brian-brazil) [inline discovery/triton/triton.go]: The way we handle arrays is to have a single label where they're joined with commas, and also surrounded with commas.

File: discovery/triton/triton.go (hunk 1 above)
Comment (richardkiene) [reply discovery/triton/triton.go]: I assume you're talking about the `:` at the end of `json:"groups:` and the missing `"`? If so it's amazing this actually works with it like that :). I'll fix this right now.

Comment (richardkiene) [reply discovery/triton/triton.go]: Fixed

File: discovery/triton/triton.go (hunk 2 above)
Comment (richardkiene) [reply discovery/triton/triton.go]: I added escaping anyway, so this should be fixed.

File: discovery/triton/triton.go (hunk 3 above)
Comment (richardkiene) [reply discovery/triton/triton.go]: done

**Comment by richardkiene** [comment]
@brian-brazil I think I've fixed everything you pointed out, thanks.

File: discovery/triton/triton.go [hunk 4]
```diff
@@ -187,6 +192,11 @@ func (d *Discovery) refresh() (tg *targetgroup.Group, err error) {
 	}()
 
 	var endpoint = fmt.Sprintf("https://%s:%d/v%d/discover", d.sdConfig.Endpoint, d.sdConfig.Port, d.sdConfig.Version)
+	if d.sdConfig.Groups != nil {
+		groups := html.EscapeString(strings.Join(d.sdConfig.Groups, ","))
```
Comment (brian-brazil) [inline discovery/triton/triton.go]: url.QueryEscape

File: discovery/triton/triton.go [hunk 5]
```diff
@@ -187,6 +192,11 @@ func (d *Discovery) refresh() (tg *targetgroup.Group, err error) {
 	}()
 
 	var endpoint = fmt.Sprintf("https://%s:%d/v%d/discover", d.sdConfig.Endpoint, d.sdConfig.Port, d.sdConfig.Version)
+	if d.sdConfig.Groups != nil {
```
Comment (brian-brazil) [inline discovery/triton/triton.go]: `len(d.sdConfig.Groups) > 0` would be more typical

File: discovery/triton/triton_test.go [hunk 6]
```diff
@@ -112,6 +136,7 @@ func TestTritonSDRefreshMultipleTargets(t *testing.T) {
 	var (
 		dstr = `{"containers":[
 		 	{
+                                "groups":["foo","bar","baz"],
```
Comment (brian-brazil) [inline discovery/triton/triton_test.go]: indentation is off

File: discovery/triton/triton.go [hunk 7]
```diff
@@ -219,6 +229,12 @@ func (d *Discovery) refresh() (tg *targetgroup.Group, err error) {
 		}
 		addr := fmt.Sprintf("%s.%s:%d", container.VMUUID, d.sdConfig.DNSSuffix, d.sdConfig.Port)
 		labels[model.AddressLabel] = model.LabelValue(addr)
+
+		if container.Groups != nil && len(container.Groups) > 0 {
+			name := strings.Join(container.Groups, ",")
+			labels[tritonLabelGroups+"_"+model.LabelName(","+name+",")] = model.LabelValue(name)
```
Comment (brian-brazil) [inline discovery/triton/triton.go]: The name should be fixed, look at how Consul tags are handled

File: discovery/triton/triton_test.go (hunk 6 above)
Comment (richardkiene) [reply discovery/triton/triton_test.go]: This appears to "look" incorrect because of the way GitHub formats it. I double checked the source code, re-ran `go fmt`, and also ran `make style` without issue.

File: discovery/triton/triton.go (hunk 4 above)
Comment (richardkiene) [reply discovery/triton/triton.go]: Will do

File: discovery/triton/triton.go (hunk 5 above)
Comment (richardkiene) [reply discovery/triton/triton.go]: Will do

File: discovery/triton/triton.go (hunk 7 above)
Comment (richardkiene) [reply discovery/triton/triton.go]: Okay, I understand now. Fixing this currently.

File: docs/configuration/configuration.md [hunk 8]
```diff
@@ -947,6 +949,14 @@ dns_suffix: <string>
 # often the same value as dns_suffix.
 endpoint: <string>
 
+# The group(s) to filter discovery results by.
```
Comment (brian-brazil) [inline docs/configuration/configuration.md]: This is a bit wordy, Consul has "A list of services for which targets are retrieved. If omitted, all services
 are scraped."

Comment (richardkiene) [reply docs/configuration/configuration.md]: Done.

**Comment by richardkiene** [comment]
@brian-brazil I've updated the commit message to reflect the additional changes. I also addressed your concern about the documentation being to wordy regarding the groups configuration option. I will ping you again once CMON support for this change has landed in master.

Thanks for the review!

**Comment by richardkiene** [comment]
@brian-brazil the corresponding changes to triton-cmon have been [merged](https://github.com/joyent/triton-cmon/commit/f837f3e96bdc0e47ba2c3ad063b244a8092b06c0) into master. Thanks!

**Comment by brian-brazil** [comment]
Thanks!
